File: app/src/main/java/in/digistorm/aksharam/LetterCategoryAdapter.java
package in.digistorm.aksharam;

/*
 * Copyright (c) 2022 Alan M Varghese <alan@digistorm.in>
 *
 * This files is part of Aksharam, a script teaching app for Indic
 * languages.
 *
 * Aksharam is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Aksharam is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even teh implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import android.content.Context;
import android.graphics.Point;
import android.graphics.Typeface;
import android.os.Build;
import android.util.Log;
import android.util.TypedValue;
import android.view.Display;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseExpandableListAdapter;
import androidx.gridlayout.widget.GridLayout;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.RequiresApi;
import androidx.fragment.app.FragmentActivity;

public class LetterCategoryAdapter extends BaseExpandableListAdapter {
    private final FragmentActivity activity;
    private final String logTag = "LetterCategoryAdapter";

    private final String[] headers;

    // The adapters copy of the parent fragment
    private LettersTabFragment lettersTabFragment;

    public LetterCategoryAdapter(FragmentActivity activity, LettersTabFragment ltf) {
        lettersTabFragment = ltf;
        headers = lettersTabFragment.getTransliterator().getLangDataReader()
                .getCategories().keySet().toArray(new String[0]);
        this.activity = activity;
    }

    @Override
    public int getGroupCount() {
        return lettersTabFragment.getTransliterator().getLangDataReader()
                .getCategories().size();
    }

    @Override
    public int getChildrenCount(int groupPosition) {
        // Each category has a single child - a list of all letters in that category
        return 1;
    }

    @Override
    public Object getGroup(int groupPosition) {
        return headers[groupPosition];
    }

    @Override
    public Object getChild(int groupPosition, int childPosition) {
        return lettersTabFragment.getTransliterator().getLangDataReader()
                .getCategories().get(headers[groupPosition]);
    }

    @Override
    public long getGroupId(int groupPosition) {
        return groupPosition;
    }

    @Override
    public long getChildId(int groupPosition, int childPosition) {
        return childPosition;
    }

    @Override
    public boolean hasStableIds() {
        return false;
    }

    @Override
    public View getGroupView(int groupPosition, boolean isExpanded, View convertView,
                             ViewGroup parent) {
        Log.d(logTag, "getting groupview for position " + groupPosition);
        if (convertView == null) {
            LayoutInflater layoutInflater = (LayoutInflater) this.activity
                    .getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            convertView = layoutInflater.inflate(R.layout.letter_category_header, null);
        }

        TextView letterCategoryHeaderTV = convertView.findViewById(R.id.LetterCategoryHeaderText);
        // Set some padding on the left so that the text does not overwrite the expand indicator
        letterCategoryHeaderTV.setPadding(100,
                letterCategoryHeaderTV.getPaddingTop(),
                letterCategoryHeaderTV.getPaddingRight(),
                letterCategoryHeaderTV.getPaddingBottom());
        letterCategoryHeaderTV.setTypeface(null, Typeface.BOLD);
        letterCategoryHeaderTV.setText(headers[groupPosition].toUpperCase());
        letterCategoryHeaderTV.setTextSize(TypedValue.COMPLEX_UNIT_SP, 25);

        return convertView;
    }

    @RequiresApi(api = Build.VERSION_CODES.M)
    @Override
    public View getChildView(int groupPosition, int childPosition, boolean isLastChild,
                             View convertView, ViewGroup parent) {
        if (convertView == null) {
            LayoutInflater layoutInflater = (LayoutInflater) this.activity
                    .getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            convertView = layoutInflater.inflate(R.layout.letter_category_content, null);
        }
        Log.d(logTag, "creating grid for group: " + groupPosition);
        Log.d(logTag, "group is: " + lettersTabFragment.getTransliterator()
                .getLangDataReader().getCategories().get(headers[groupPosition]));
        GridLayout gridLayout = convertView.findViewById(R.id.LetterGrid);
        gridLayout.removeAllViews();
        gridLayout.setClickable(true);

        String[] letters = lettersTabFragment.getTransliterator().getLangDataReader()
                .getCategories().get(headers[groupPosition]).toArray(new String[0]);
        for(String letter: letters) {
            LinearLayout linearLayout = new LinearLayout(activity);
            linearLayout.setOrientation(LinearLayout.VERTICAL);
            Display display = activity.getWindowManager().getDefaultDisplay();
            Point size = new Point();
            try {
                display.getRealSize(size);
            } catch (NoSuchMethodError err) {
                display.getSize(size);
            }
            LinearLayout.LayoutParams layoutParams = new LinearLayout.LayoutParams(
                    size.x/6,
                    LinearLayout.LayoutParams.WRAP_CONTENT);
            linearLayout.setLayoutParams(layoutParams);
            linearLayout.setGravity(Gravity.CENTER);

            TextView tv = new TextView(activity);
            tv.setText(letter);
            ViewGroup.MarginLayoutParams tvLayoutParams = new ViewGroup.MarginLayoutParams(
                    ViewGroup.LayoutParams.WRAP_CONTENT,
                    ViewGroup.LayoutParams.WRAP_CONTENT);

            int pixels = lettersTabFragment.getResources().getDimensionPixelSize(R.dimen.letter_grid_margin);
            tvLayoutParams.setMargins(pixels, pixels, pixels, pixels);
            tv.setLayoutParams(tvLayoutParams);
            pixels = lettersTabFragment.getResources().getDimensionPixelSize(R.dimen.letter_grid_padding);
            tv.setPadding(pixels, pixels, pixels, pixels);
            tv.setTextSize(TypedValue.COMPLEX_UNIT_SP, 22);

            linearLayout.setClickable(true);
            linearLayout.setLongClickable(true);
            linearLayout.setOnLongClickListener(v -> {
                Log.d(logTag, letter + " long clicked!");

                LetterInfoFragment letterInfoFragment = LetterInfoFragment.newInstance(letter, lettersTabFragment);
                MainActivity.replaceTabFragment(0, letterInfoFragment);
                return true;
            });

            linearLayout.setOnClickListener(v -> {
                Log.d(logTag, letter + " clicked!");
                if (tv.getText().toString().equals(letter)) {
                    if (!lettersTabFragment.getLettersTabFragmentLanguage().equalsIgnoreCase(
                            lettersTabFragment.getLettersTabFragmentTargetLanguage()))
                        tv.setText(lettersTabFragment.getTransliterator().transliterate(
                                letter,
                                lettersTabFragment.getLettersTabFragmentTargetLanguage()));
                    else
                        Log.d(logTag, "source lang = target lang");
                }
                else
                    tv.setText(letter);
            });

            linearLayout.addView(tv);
            gridLayout.addView(linearLayout);
        }
        return convertView;
    }

    @Override
    public boolean isChildSelectable(int groupPosition, int childPosition) {
        return false;
    }
}

File: app/src/main/kotlin/in/digistorm/aksharam/activities/main/fragments/letters/LetterCategoryAdapter.kt
/*
 * Copyright (c) 2022-2025 Alan M Varghese <alan@digistorm.in>
 *
 * This files is part of Aksharam, a script teaching app for Indic
 * languages.
 *
 * Aksharam is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Aksharam is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even teh implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package `in`.digistorm.aksharam.activities.main.fragments.letters

import android.animation.ObjectAnimator
import android.util.SparseBooleanArray
import `in`.digistorm.aksharam.R
import `in`.digistorm.aksharam.databinding.LetterCategoryBinding
import `in`.digistorm.aksharam.activities.main.util.logDebug
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AccelerateDecelerateInterpolator
import android.widget.TextView
import androidx.core.util.getOrDefault
import androidx.core.util.set
import androidx.core.view.ViewCompat
import androidx.gridlayout.widget.GridLayout
import androidx.navigation.NavDirections
import androidx.navigation.findNavController
import androidx.navigation.fragment.FragmentNavigatorExtras
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import androidx.transition.ChangeBounds
import androidx.transition.Transition
import androidx.transition.TransitionManager
import com.google.android.material.card.MaterialCardView
import `in`.digistorm.aksharam.activities.main.helpers.upperCaseFirstLetter
import `in`.digistorm.aksharam.activities.main.language.Category

/* RecyclerView adapter for each category of letters (e.g. Vowels, Consonants, etc).
   Each category is displayed in a separate Material 3 Card View.
 */
class LetterCategoryAdapter(
    private var letterOnLongClickAction: (String) -> NavDirections,
    private val cardStates: SparseBooleanArray = SparseBooleanArray()
): ListAdapter<Category, LetterCategoryAdapter.LetterCategoryCardViewHolder>(
    LetterCategoryDiff { cardStates }
) {
    private val logTag = this.javaClass.simpleName

    init {
        logDebug(logTag, "LetterCategoryAdapter initialised with list: $currentList")
    }

    private class LetterCategoryDiff(val getCardStates: (() -> SparseBooleanArray)): DiffUtil.ItemCallback<Category>() {
        private val logTag = this.javaClass.simpleName

        override fun areItemsTheSame(
            oldItem: Category,
            newItem: Category
        ): Boolean {
            logDebug(logTag, "Are oldItem: $oldItem and newItem: $newItem the same:" +
                    " ${oldItem == newItem}")
            /**
             * Items are same if:
             * 1. Their names are same,
             * 2. For each value in the Arraylist, first of the pair are the same
             *    This means only the transliteration language has been changed.
             **/
            if(oldItem.name == newItem.name) {
                oldItem.letterPairs.forEachIndexed { i, letterPair ->
                    if(letterPair.first != newItem.letterPairs[i].first) {
                        getCardStates.invoke().clear()
                        return false
                    }
                }
            }
            else
                return false
            return true
        }

        // Iff items are same, the system checks whether their contents are also same
        override fun areContentsTheSame(
            oldItem: Category,
            newItem: Category
        ): Boolean {
            logDebug(logTag, "Are contents of oldItem: $oldItem and " +
                    "newItem: $newItem the same: false")
            return false
        }
    }

    inner class LetterCategoryCardViewHolder(
        private val letterCategoryBinding: LetterCategoryBinding
    ): RecyclerView.ViewHolder(letterCategoryBinding.root) {
        // private val logTag = this.javaClass.simpleName

        private var position: Int = RecyclerView.NO_POSITION

        private val transition: Transition
            get() {
                return ChangeBounds()
                    .setDuration(300)
                    .setInterpolator(AccelerateDecelerateInterpolator())
            }

        val cardView: MaterialCardView by lazy {
            letterCategoryBinding.cardView
        }

        val letterGrid: GridLayout by lazy {
            letterCategoryBinding.letterGrid
        }

        fun getTag(): Any {
            return cardView.tag
        }

        private fun isCollapsed(): Boolean {
            return cardStates.getOrDefault(position, false)
        }

        var text: CharSequence
            get() = letterCategoryBinding.letterCategoryHeaderText.text
            set(value) {
                letterCategoryBinding.letterCategoryHeaderText.text = value
            }

        var findNextSiblings: (() -> MutableList<MaterialCardView>)? = null

        fun initialise(position: Int) {
            this.position = position

            if(isCollapsed()) {
                letterGrid.visibility = View.GONE
            } else {
                letterGrid.visibility = View.VISIBLE
            }

            letterCategoryBinding.apply {
                letterCategoryHeader.setOnClickListener {
                    val logTag = "categoryHeaderClickListener"
                    val imageView = downArrow

                    if (isCollapsed()) {
                        TransitionManager.beginDelayedTransition(
                            cardView.parent as ViewGroup,
                            transition.addTarget(cardView)
                        )
                        letterGrid.visibility = View.VISIBLE

                        ObjectAnimator.ofFloat(imageView, "rotation", 0f).apply {
                            duration = 300
                            start()
                        }
                        cardStates[position] = false
                    } else {
                        logDebug(logTag, "Finding siblings...")
                        val siblings: MutableList<MaterialCardView> = findNextSiblings?.invoke()!!
                        val newTransition: Transition = transition.apply {
                            logDebug(logTag, "adding siblings as target for animation")
                            logDebug(logTag, "Siblings size: ${siblings.size}")
                            siblings.forEach { sibling ->
                                logDebug(logTag, "Found CardView for: " +
                                        "${sibling.findViewById<TextView>(R.id.letter_category_header_text)?.text}")
                                addTarget(sibling)
                            }
                        }
                        TransitionManager.beginDelayedTransition(cardView.parent as ViewGroup, newTransition)
                        letterGrid.visibility = View.GONE
                        ObjectAnimator.ofFloat(imageView, "rotation", 180f).apply {
                            duration = 300
                            start()
                        }
                        cardStates[position] = true
                    }
                }
            }

        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): LetterCategoryCardViewHolder {
        val letterCategoryBinding = LetterCategoryBinding.inflate(
            LayoutInflater.from(parent.context), parent, false
        )
        return LetterCategoryCardViewHolder(letterCategoryBinding)
    }

    override fun onBindViewHolder(holder: LetterCategoryCardViewHolder, position: Int) {
        logDebug(logTag, "Binding view for position: $position")
        logDebug(logTag, "Item: ${getItem(position)}")
        holder.cardView.tag = getItem(position).name
        holder.initialise(position)

        // Give an ExpandableCardView the ability to find its sibling, so that they are animated correctly
        holder.findNextSiblings = fun () : MutableList<MaterialCardView> {
            var pos = holder.bindingAdapterPosition
            return if(pos == RecyclerView.NO_POSITION) {
                logDebug(logTag, "bindingAdapterPosition was NO_POSITION")
                mutableListOf()
            } else {
                val siblingList = mutableListOf<MaterialCardView>()
                logDebug(logTag, "bindingAdapterPosition is $pos")
                while(++pos < itemCount) {
                    (((holder.cardView.parent as? RecyclerView)
                        ?.findViewHolderForAdapterPosition(pos)) as? LetterCategoryCardViewHolder)?.cardView?.let {
                        siblingList.add(it)
                    }
                }
                mutableListOf()
            }
        }

        holder.text = getItem(position).name.upperCaseFirstLetter()

        initialiseLetterGrid(holder, position)
    }

    /* Initialises the LetterGrid for a category of letters (i.e, this is run once for every
       category.) */
    private fun initialiseLetterGrid(
        letterCategoryCardViewHolder: LetterCategoryCardViewHolder,
        position: Int,
    ) {
        letterCategoryCardViewHolder.letterGrid.apply {
            removeAllViews()

            logDebug(logTag, "Initialising letter grid for position: $position")
            val category = getItem(position).name
            logDebug(logTag, "Current category: $category")
            logDebug(logTag, "Letter Pairs: ${getItem(position).letterPairs}")
            for(letterPair in getItem(position).letterPairs) {
                val letterView: LetterPairView = LayoutInflater
                    .from(letterCategoryCardViewHolder.letterGrid.context)
                    .inflate(R.layout.letter_pair_view,
                        letterCategoryCardViewHolder.letterGrid, false) as LetterPairView
                letterView.setOnLongClickListener { view ->
                    val extras = FragmentNavigatorExtras(letterView to "${letterPair.first}_heading")

                    view.findNavController().navigate(letterOnLongClickAction(letterPair.first), extras)
                    true
                }
                letterView.letters = letterPair
                letterView.tag = letterPair.first
                ViewCompat.setTransitionName(letterView, "${letterPair.first}_letter")
                letterCategoryCardViewHolder.letterGrid.addView(letterView)
            }
        }
    }
}
